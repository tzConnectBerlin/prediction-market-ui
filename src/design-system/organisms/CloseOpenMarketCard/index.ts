export { CloseOpenMarketCard, CloseOpenMarketCard as default } from './CloseOpenMarketCard';
export type { CloseOpenMarketProps } from './CloseOpenMarketCard';
