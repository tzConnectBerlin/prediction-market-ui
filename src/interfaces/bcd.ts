export interface SimilarContract {
  added?: number;
  address: string;
  alias?: string;
  annotations?: string[];
  balance?: number;
  count?: number;
  delegate?: string;
  delegate_alias?: string;
  entrypoints?: string[];
  fail_strings?: string[];
  found_by?: string;
  hardcoded?: string[];
  hash: string;
  id?: string;
  language?: string;
  last_action?: string;
  level?: number;
  manager?: string;
  migrations_count?: number;
  network?: string;
  project_id?: string;
  removed?: number;
  slug?: string;
  subscription?: Subscription;
  tags?: string[];
  timestamp?: string;
  tokens?: TokenBalance[];
  total_subscribed?: number;
  total_withdrawn?: number;
  tx_count?: number;
  verification_source?: string;
  verified?: boolean;
}

export interface Subscription {
  address?: string;
  alias?: string;
  network?: string;
  sentry_dsn?: string;
  sentry_enabled?: boolean;
  subscribed_at?: string;
  watch_calls?: boolean;
  watch_deployments?: boolean;
  watch_errors?: boolean;
  watch_mempool?: boolean;
  watch_migrations?: boolean;
  watch_same?: boolean;
  watch_similar?: boolean;
}

export interface TokenBalance {
  balance?: number;
  contract?: string;
  decimals?: number;
  name?: string;
  symbol?: string;
  token_id?: number;
}

export interface SimilarContractResponse {
  contracts?: SimilarContract[];
  count?: number;
}
