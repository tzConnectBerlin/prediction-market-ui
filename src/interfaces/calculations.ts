export interface ClosePositionReturn {
  aToSwap: number;
  aLeft: number;
  bReceived: number;
  aToSwapWithSlippage: number;
}

export interface ClosePositionBothReturn
  extends Omit<ClosePositionReturn, 'bReceived' | 'aToSwapWithSlippage'> {
  bHeld: number;
}

export interface BuyPosition {
  price: number;
  swap: number;
  quantity: number;
}
