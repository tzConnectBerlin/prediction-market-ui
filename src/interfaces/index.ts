export * from './wallet';
export * from './router';
