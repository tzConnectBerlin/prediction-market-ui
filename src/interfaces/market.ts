import { MichelsonMap } from '@taquito/taquito';
import { BigNumber } from 'bignumber.js';

export type QuestionType = string;

export interface Links {
  label: string;
  url: string;
}

export enum MarketStateType {
  auctionRunning = 'auction',
  marketBootstrapped = 'market',
}

export enum Role {
  participant = 'Participant',
  adjudicator = 'Adjudicator',
}

export interface AuctionNode {
  auctionRunningAuctionPeriodEnd: string;
  auctionRunningQuantity: string;
  auctionRunningYesPreference: string;
  auctionRunningUniswapContribution: string;
}

export interface MarketNode {
  auctionRewardCurrencyPool: string;
  liquidityRewardPool: string;
  marketCurrencyPool: string;
  bootstrapYesProbability: string;
  liquidityRewardSupplyUpdatedAtBlock: string;
  winningPrediction?: string;
  resolutionResolvedAtBlock?: number;
  marketBootstrappedBootstrappedAtBlock: string;
}

export interface PortfolioMarket {
  question: (string | undefined)[];
  holdings: string[] | string;
  price: string[] | string;
  total: string[] | string;
}

export interface PortfolioAuction {
  question: string;
  probability: string;
  quantity: string;
}

export interface MarketStateNode {
  node: AuctionNode | MarketNode;
}

export interface StorageMarketMapMarketBootstrappeds {
  nodes: MarketStateNode[];
}

export interface StorageMarketMapAuctionRunnings {
  nodes: MarketStateNode[];
}

export interface GraphMarketNode {
  node: GraphMarket;
}
export interface GraphMarket {
  id: number;
  block: number;
  deleted: boolean;
  marketId: string;
  metadataIpfsHash: string;
  metadataDescription: string;
  metadataAdjudicator: string;
  state: string;
  dateTime: {
    bakedAt: string;
  };
  storageMarketMapAuctionRunnings: StorageMarketMapAuctionRunnings;
  storageMarketMapMarketBootstrappeds: StorageMarketMapMarketBootstrappeds;
}

export interface IPFSMarketData {
  question: QuestionType;
  auctionEndDate: string;
  iconURL?: string;
  ticker: string;
}

export interface Market extends Partial<AuctionNode>, Partial<MarketNode>, IPFSMarketData {
  marketId: string;
  ipfsHash: string;
  description: string;
  adjudicator: string;
  state: MarketStateType;
  yesPrice: number;
  liquidity?: number | string;
  block: number;
  bakedAt: string;
}

export interface AllMarkets {
  markets: {
    marketNodes: GraphMarket[];
  };
}

export interface AllLedgers {
  ledgers: {
    ledgerMaps: LedgerMap[];
  };
}

export interface AllMarketsLedgers {
  markets: {
    marketNodes: GraphMarket[];
  };
  ledgers: {
    ledgerMaps: LedgerMap[];
  };
}

export interface MarketCardStatistic {
  type: string;
  value: string | number;
  changes?: string;
  currency?: Currency;
  tokenType?: TokenType;
}

export interface MarketCardToken {
  type: TokenType;
  value: number | string;
}

export interface MarketCardData extends Market {
  tokens?: MarketCardToken[];
  statistics?: MarketCardStatistic[];
}

export interface CreateMarket {
  marketId: number;
  ipfsHash: string;
  description: string;
  adjudicator: string;
  auctionEnd: string;
  initialBid: number;
  initialContribution: number;
  tokenType: 'fa12' | 'fa2';
  tokenAddress: string;
}

export interface GraphBet {
  probability: string;
  quantity: string;
}

export interface Bet {
  block: number;
  originator: string;
  probability: number;
  quantity: number;
  marketId: string;
}

export interface BetEdge {
  bet: GraphBet;
}

export interface GraphBets {
  totalBets: number;
  betEdges: BetEdge[];
}

export interface LqtProviderNode {
  id: number;
  marketId: string;
  originator: string;
  bets: GraphBets;
  block: number;
}

export interface LqtProviderEdge {
  lqtProviderNode: LqtProviderNode;
}

export interface StorageLiquidityProviderMaps {
  lqtProviderEdge: LqtProviderEdge[];
}

export interface AllBets {
  storageLiquidityProviderMaps: StorageLiquidityProviderMaps;
}

export enum MarketTradeType {
  payIn = 'payIn',
  payOut = 'payOut',
}

export enum FormType {
  buy = 'Buy',
  sell = 'Sell',
  addLiquidity = 'Add Liquidity',
  removeLiquidity = 'Remove Liquidity',
}

export interface LiquidityValues {
  probability: number;
  quantity: number;
}

export interface TokenSupplyMap {
  id: number;
  tokenId: string;
  totalSupply: string;
  tokenReserve: string;
  deleted: boolean;
}

export interface StorageSupplyMaps {
  supplyMaps: TokenSupplyMap[];
}

export interface AllTokens {
  storageSupplyMaps: StorageSupplyMaps;
}

export enum TokenType {
  yes = 'Yes',
  no = 'No',
}

export interface LedgerMap {
  id: number;
  block: number;
  deleted: boolean;
  owner: string;
  tokenId: string;
  quantity: string;
}

export interface StorageLedgerMaps {
  ledgerMaps: LedgerMap[];
}

export interface Token {
  id: number;
  tokenId: string;
  quantity: string;
  block: number;
  dateTime: {
    bakedAt: string;
  };
}

export interface TokenQuantity {
  token: Token[];
}

export interface MarketPricePoint {
  yesPrice: number;
  block: number;
  bakedAt: string;
}

export interface AddressTokens {
  tokenQuantity: TokenQuantity;
}

export interface AuctionMarkets {
  [key: string]: Market[];
}

// TODO: clean the stuff below

export interface QuestionState {
  [key: number]: {
    [key in MarketStateType]: symbol;
  };
}

export enum Currency {
  USD = '$',
  EUR = 'â‚¬',
}

export type CurrencyTypes = keyof typeof Currency;

export enum MarketEntrypoint {
  Bid = 'bid',
  BuyYes = 'buyYes',
  BuyNo = 'buyNo',
  ClaimWinnings = 'claimWinnings',
  CloseAuction = 'closeAuction',
  CreateQuestion = 'createQuestion',
  WithdrawAuction = 'withdrawAuction',
}

export interface CreateQuestion {
  question: QuestionType;
  auctionEndDate: Date | string;
  marketCloseDate: Date | string;
  iconURL?: string;
  yesAnswer: string;
  rate?: number;
  quantity?: number;
}

export interface Bid {
  question: QuestionType;
  rate: number;
  quantity: number;
}

export interface BuyToken {
  tokenType: TokenType;
  question: QuestionType;
  quantity: number;
}

export interface CloseMarket {
  tokenType: TokenType;
  question: QuestionType;
}

export interface ClaimWinnings {
  question: QuestionType;
}

export interface QuestionMetaData extends CreateQuestion {
  hash: string;
}

export interface BidEntry {
  rate: number;
  quantity: number;
  total_token: number;
}

export type BidRegistry = MichelsonMap<string, BidEntry>;

export interface QuestionEntry {
  owner: string;
  state: QuestionState;
  auction_end: string;
  market_close: string;
  auction_bids: BidRegistry;
  uniswap_pool?: unknown;
  tokens: {
    yes_token_id: BigNumber;
    no_token_id: BigNumber;
    lqt_token_id: BigNumber;
  };
  answer?: string | null;
  winning_token?: BigNumber | null; // token that will receive the reward
  total_auction_quantity: BigNumber;
  yes_preference: BigNumber;
  uniswap_contribution_factor: BigNumber;
}

export interface AuctionData {
  yes: number;
  no: number;
  participants: number;
}

export interface AuctionDataMap {
  [key: string]: AuctionData;
}

export interface MarketFilter {
  auctions: boolean;
  allMarkets: boolean;
  openMarkets: boolean;
  closedMarkets: boolean;
  onlyMyMarkets: boolean;
}

/**
 * Temp middleware interfaces
 */

export interface BidEntryMDW {
  rate: string;
  quantity: string;
  total_token: string;
}

export interface BidRegistryMDW {
  [key: string]: BidEntryMDW;
}

export interface QuestionEntryMDW {
  owner: string;
  state: {
    [key in MarketStateType]?: null;
  };
  auction_end: string;
  market_close: string;
  auction_bids: BidRegistryMDW;
  uniswap_pool?: unknown;
  tokens: {
    yes_token_id: string;
    no_token_id: string;
    lqt_token_id: string;
  };
  answer?: string | null;
  winning_token?: string | null;
  total_auction_quantity: string;
  yes_preference: string;
  uniswap_contribution_factor: string;
  price_yes?: string;
}

export interface QuestionEntryMDWMap {
  [key: string]: QuestionEntryMDW;
}

export interface StableCoinResponse {
  [key: string]: string;
}

export interface LedgerBalanceResponse {
  [tokenId: string]: { [address: string]: string };
}

export interface ContractError {
  [key: number]: string;
}

export interface DropDownItems {
  label: string;
  value: string | number;
}
