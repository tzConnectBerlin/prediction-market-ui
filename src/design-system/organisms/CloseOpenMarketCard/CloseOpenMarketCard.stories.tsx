import React from 'react';
import { ToastProvider } from 'react-toast-notifications';
import { WalletProvider } from '@tezos-contrib/react-wallet-provider';
import { Story, Meta } from '@storybook/react/types-6-0';
import { CloseOpenMarketCard, CloseOpenMarketProps } from './CloseOpenMarketCard';
import { MarketStateType } from '../../../interfaces';

export default {
  title: 'Organism/CloseOpenMarketCard',
  component: CloseOpenMarketCard,
} as Meta;

const Template: Story<CloseOpenMarketProps> = (args) => (
  <WalletProvider name="Temple">
    <ToastProvider>
      <CloseOpenMarketCard {...args} />
    </ToastProvider>
  </WalletProvider>
);

export const Default = Template.bind({});

export const Resolved = Template.bind({});
Resolved.args = {
  marketId: '1',
  marketPhase: MarketStateType.marketBootstrapped,
  winningPrediction: 'yes',
};
