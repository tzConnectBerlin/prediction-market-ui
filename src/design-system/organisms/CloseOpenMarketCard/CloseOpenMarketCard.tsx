import * as React from 'react';
import { Card, CardContent, useTheme } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import styled from '@emotion/styled';
import { MarketStateType } from '../../../interfaces';
import Button from '../../atoms/Button';
import { Typography } from '../../atoms/Typography';
import { ResolveMarketModal } from '../ResolveMarketModal';

export interface CloseOpenMarketProps {
  marketId: string;
  marketPhase: MarketStateType;
  adjudicator?: string;
  address?: string;
  winningPrediction?: string;
  handleCloseAuction?: (id: string) => Promise<void>;
  handleResolveMarket?: (id: string) => Promise<void>;
  closeMarketId?: string;
  setCloseMarketId?: React.Dispatch<React.SetStateAction<string>>;
  auctionParticipant?: boolean;
}

const StyledCard = styled(Card)`
  margin-bottom: 1.5rem;
  padding: 1rem;
  padding-bottom: 0.5rem;
`;

const StyledDiv = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
`;
export const CloseOpenMarketCard: React.FC<CloseOpenMarketProps> = ({
  adjudicator,
  address,
  winningPrediction,
  marketPhase,
  marketId,
  handleCloseAuction,
  handleResolveMarket,
  closeMarketId,
  setCloseMarketId,
  auctionParticipant,
}) => {
  const theme = useTheme();
  const handleClose = () => setCloseMarketId && setCloseMarketId('');
  const { t } = useTranslation('common');

  return marketPhase === MarketStateType.marketBootstrapped &&
    adjudicator !== address &&
    !winningPrediction ? null : (
    <StyledCard>
      {handleResolveMarket && auctionParticipant && (
        <ResolveMarketModal
          open={!!closeMarketId}
          handleClose={handleClose}
          handleSubmit={() => handleResolveMarket(marketId)}
        />
      )}
      <CardContent>
        {marketPhase === 'auction' && (
          <Typography marginBottom="1.5rem" size="14px">
            {t('closeAuction')}
          </Typography>
        )}
        {winningPrediction && (
          <StyledDiv>
            <Typography>{t('resolution')}</Typography>
            <Typography
              color={
                winningPrediction === 'yes' ? theme.palette.success.main : theme.palette.error.main
              }
            >
              {winningPrediction.charAt(0).toUpperCase() + winningPrediction.slice(1)}
            </Typography>
          </StyledDiv>
        )}
        {(!winningPrediction && adjudicator === address) ||
          (marketPhase === 'auction' && auctionParticipant && (
            <Button
              fullWidth
              label={marketPhase === 'auction' ? t('openMarketToTrade') : t('closeMarket')}
              onClick={
                marketPhase === 'auction'
                  ? () => handleCloseAuction && handleCloseAuction(marketId)
                  : () => setCloseMarketId && setCloseMarketId(marketId)
              }
            />
          ))}
      </CardContent>
    </StyledCard>
  );
};
