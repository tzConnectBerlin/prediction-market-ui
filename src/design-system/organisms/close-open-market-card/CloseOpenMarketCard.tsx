import React from 'react';
import { Card, CardContent, useTheme } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import styled from '@emotion/styled';
import { useWallet } from '@tezos-contrib/react-wallet-provider';
import { useToasts } from 'react-toast-notifications';
import { MarketTradeType, Token } from '../../../interfaces';
import { MarketPosition, MarketPositionProps } from '../../molecules/MarketPosition';
import { claimWinnings, closeAuction } from '../../../contracts/Market';
import { logError } from '../../../logger/logger';
import Button from '../../atoms/Button';

export interface CloseOpenMarketProps {
  tokenList: Token[];
  outcomeItems: string[];
  marketId: string;
  connected: boolean;
}
export const CloseOpenMarketCard: React.FC<CloseOpenMarketProps> = ({
  tokenList,
  outcomeItems,
  marketId,
  connected,
}) => {
  const { addToast } = useToasts();
  const [, setCloseMarketId] = React.useState('');
  const handleClose = () => setCloseMarketId('');
  const theme = useTheme();
  const { activeAccount } = useWallet();
  const { t } = useTranslation('common');
  const [value, setValue] = React.useState(0);

  const handleChange = (event: React.SyntheticEvent, newValue: number) => {
    setValue(newValue);
  };
  const handleCloseAuction = React.useCallback(
    async (id: string) => {
      if (activeAccount?.address && id) {
        try {
          await closeAuction(id, true);
        } catch (error) {
          logError(error);
          const errorText = error?.data[1]?.with?.string || t('txFailed');
          addToast(errorText, {
            appearance: 'error',
            autoDismiss: true,
          });
        }
      }
    },
    [activeAccount?.address, addToast, t],
  );
  const handleClaimWinnings = React.useCallback(
    async (handleId: string) => {
      if (activeAccount?.address && handleId) {
        try {
          const hash = await claimWinnings(handleId);
          if (hash) {
            handleClose();
          }
        } catch (error) {
          logError(error);
          const errorText = error?.data[1]?.with?.string || t('txFailed');
          addToast(errorText, {
            appearance: 'error',
            autoDismiss: true,
          });
        }
      }
    },
    [activeAccount?.address, addToast, t],
  );

  return (
    <Card>
      <CardContent>
        {!outcomeItems.length && (
          <Button
            style={{ width: '100%' }}
            label={t('claimWinningsPage')}
            onClick={() => handleClaimWinnings(marketId)}
          />
        )}
        {!outcomeItems.length && (
          <Button
            style={{ width: '100%' }}
            label={t('claimWinningsPage')}
            onClick={() => handleClaimWinnings(marketId)}
          />
        )}
      </CardContent>
    </Card>
  );
};
