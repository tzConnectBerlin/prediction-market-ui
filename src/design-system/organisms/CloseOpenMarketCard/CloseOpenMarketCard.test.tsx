import renderer from 'react-test-renderer';
import { CloseOpenMarketCard } from './CloseOpenMarketCard';
import { MarketStateType } from '../../../interfaces';

const defaultArgs = {
  adjudicator: 'tim',
  marketId: '1',
  marketPhase: MarketStateType.auctionRunning,
};

const WrappedComponent: React.FC<any> = (args: any) => <CloseOpenMarketCard {...args} />;

describe('Snapshot - render TradeForm', () => {
  it('renders correctly with default props', () => {
    const BuyForm = renderer.create(<WrappedComponent {...defaultArgs} />).toJSON();
    expect(BuyForm).toMatchSnapshot();
  });
});
