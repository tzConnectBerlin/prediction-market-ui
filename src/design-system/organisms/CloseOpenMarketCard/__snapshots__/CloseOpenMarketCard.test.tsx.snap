// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot - render TradeForm renders correctly with default props 1`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  overflow: hidden;
  margin-bottom: 1.5rem;
  padding: 1rem;
  padding-bottom: 0.5rem;
}

.emotion-1 {
  padding: 16px;
}

.emotion-1:last-child {
  padding-bottom: 24px;
}

<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root emotion-0"
>
  <div
    className="MuiCardContent-root emotion-1"
  />
</div>
`;
